web/client-api/src/main/java/io/deephaven/web/client/api/lifecycle/HasLifecycle.java

```diff
@@ -3,54 +3,86 @@
  */
 package io.deephaven.web.client.api.lifecycle;
 
+import elemental2.dom.CustomEvent;
 import elemental2.dom.CustomEventInit;
+import elemental2.promise.Promise;
 import io.deephaven.web.client.api.HasEventHandling;
 import io.deephaven.web.client.api.JsTable;
 import io.deephaven.web.client.fu.JsLog;
+import io.deephaven.web.client.fu.LazyPromise;
 import io.deephaven.web.client.state.ClientTableState;
 
 import static io.deephaven.web.client.api.JsTable.EVENT_RECONNECTFAILED;
 
 /**
- * An abstraction over the lifecycle methods used to reconnect JsTable, so we can have non-JsTable bound states get the
- * same revivification treatment.
+ * An abstraction over simple lifecycle methods, to see if objects are currently connected to their server-side
+ * counterparts.
  */
-public interface HasLifecycle {
-
-    // extend HasEventHandling to get these three provided for you.
-    void suppressEvents();
-
-    void unsuppressEvents();
-
-    boolean isSuppress();
-
-    void revive(ClientTableState state);
-
+public abstract class HasLifecycle extends HasEventHandling {
     /**
-     * You should probably just call notifyDeath(this, failure), assuming you implement in an object that
-     * HasEventHandlers.
+     * Simple flag to see if already connected - that is, if true, will not fire a reconnect event unless disconnected
+     * again.
      */
-    void die(Object failure);
+    private boolean connected = true;
 
-    default void notifyDeath(HasEventHandling handler, Object error) {
+    /**
+     * Signal that reconnect was attempted and failed due to differences between client and server state.
+     */
+    public void die(Object error) {
         JsLog.debug("Die!", this, error);
         final CustomEventInit init = CustomEventInit.create();
         init.setDetail(error);
         unsuppressEvents();
-        handler.fireEvent(EVENT_RECONNECTFAILED, init);
+        fireEvent(EVENT_RECONNECTFAILED, init);
         suppressEvents();
     }
 
-    default void maybeRevive(ClientTableState state) {
-        if (isSuppress()) {
-            revive(state);
-        }
+    /**
+     * Mark the object as being connected to its corresponding server-side object.
+     */
+    public void reconnect() {
+        connected = true;
+        unsuppressEvents();
+        fireEvent(JsTable.EVENT_RECONNECT);
     }
 
-    void disconnected();
+    /**
+     * Indicate that a new session has been created on the server, and this object should re-create its corresponding
+     * server-side object if possible. Override this to implement custom behavior, being sure to call reconnect() when
+     * finished.
+     *
+     * @return a promise that will resolve when this object is reconnected
+     */
+    public Promise<?> refetch() {
```

Comment:
Is it worth making `HasLifeCycle` take a template and type this `Promise<?>`? I see if you call `JsTable.refetch()` right now it's just returning a `Promise<Object>` rather than typing correctly to `Promise<JsTable>`.

Reply:
Yeah I'll override it in table so that it throws (tables have a more complex reconnect than just "hi i'm back"), but since the ? is covariant we can override this to make it more specific in subclasses (see JsFigure and JsPartitionedTable). This type won't be visible to JS, and this method particularly isn't intended to be exposed, but I am reticent to make this type recursively generic for fear of the meaning being lost or confused in translation to TS...